pandas/core/groupby/ops.py

```diff
@@ -738,8 +738,12 @@ def agg_series(self, obj: Series, func: F):
             # TODO: can we get a performant workaround for EAs backed by ndarray?
             return self._aggregate_series_pure_python(obj, func)
 
-        elif obj.index._has_complex_internals:
-            # Preempt TypeError in _aggregate_series_fast
+        elif obj.index._has_complex_internals or obj.index.dtype == "object":
+            # (complex internals): Preempt TypeError in _aggregate_series_fast
+            # (object index dtype): _aggregate_series_fast raises TypeError
+            # when applying func because the group indiced become malformatted:
+            # correct indices are in group.index._index_data, but not in
+            # group.index._data.
```

Comment:
On master, replacing the lamba in the linked issue with `lambda x: 1` does not raise and gives the correct answer. It seems to me that just having an object dtype is not root cause here, and we will be hurting performance for this special case.

Reply:
https://github.com/pandas-dev/pandas/issues/40014#issuecomment-817196445

Reply:
I see. I will keep looking for the root cause and try to provide a fix. I will mark this PR as a draft in the meantime.

Reply:
@rhshadrach I have updated this PR.

Reply:
> 
> 
> This looks great! Can you just check on my example in the issue and report back? Do you understand why the "> 0" was causing it to fail? It's still not clear to me from the diff in this PR.

I think what happens here is that `x.gt(0)` calls `_cmp_method` and constructs a new `Series` through `_construct_result`. Note that the index length check happens when a new `Series` is constructed. `x.sum()` on the other hand does not construct a new series. The return type is `numpy.int64`. So it seems that using any function that constructs a new `Series` will fail, while the other will not. But the index state in the user-defined function is always corrupt.
A good test would be to use 

```python3
import pandas.core.common as com

def foo(x):
    com.require_length_match(x.values, x.index)
    return x
```